### PR TITLE
Port of some important fixes from devnet branch

### PR DESCRIPTION
**Pull request by paberr**
These fixes mainly concern view-changes and rebranching behaviour.

### CODE REVIEW

File: blockchain-albatross/src/blockchain.rs [hunk 1]
```diff
@@ -423,6 +434,82 @@ impl Blockchain {
...
+    fn order_chains(&self, block: &Block, prev_info: &ChainInfo, txn_option: Option<&Transaction>) -> ChainOrdering {
+        let mut chain_order = ChainOrdering::Unknown;
+        if block.parent_hash() == &self.head_hash() {
+            chain_order = ChainOrdering::Extend;
+        } else {
+            // To compare two blocks, we need to compare the view number at the intersection.
+            //   [2] - [2] - [3] - [4]
+            //      \- [3] - [3] - [3]
+            // The example above illustrates that you actually want to choose the lower chain,
+            // since it's view change happened way earlier.
```
Comment (riptl) [inline blockchain-albatross/src/blockchain.rs]:
```suggestion
            // since its view change happened way earlier.
```
:)

Comment (terorie) [inline blockchain-albatross/src/blockchain.rs]:
```suggestion
            // since its view change happened way earlier.
```

**Review by riptl — DISMISSED** [review]
I wish every PR was as clean as this.
Maybe small TODO for the future is an explanation what the BitSets in BlockDescriptor express.
It might not immediately be clear to the reader that they mark slot availability (correct term?) and also whether set membership marks a slashed slot or an active one.

File: blockchain-albatross/src/blockchain.rs (hunk 1 above)
Comment (paberr) [reply blockchain-albatross/src/blockchain.rs]: Done

**Comment by paberr** [comment]
> I wish every PR was as clean as this.

Thanks :)
> Maybe small TODO for the future is an explanation what the BitSets in BlockDescriptor express.
> It might not immediately be clear to the reader that they mark slot availability (correct term?) and also whether set membership marks a slashed slot or an active one.

True

**Review by riptl — APPROVED** [review]

**Review by jeffesquivels — APPROVED** [review]
LGTM

**Pull request merged**